runtime/src/iree/builtins/ukernel/tools/e2e_matmul_benchmark.c

```diff
@@ -0,0 +1,366 @@
+  void* rowmajor_init_out_buffer = malloc(rowmajor_out_buffer_size);
+  void* rowmajor_out_buffer = malloc(rowmajor_out_buffer_size);
+  void* packed_lhs_buffer = malloc(packed_lhs_buffer_size);
+  void* packed_rhs_buffer = malloc(packed_rhs_buffer_size);
+  void* packed_out_buffer = malloc(packed_out_buffer_size);
+  iree_uk_random_engine_t* engine = iree_uk_benchmark_random_engine(user_data);
+  // It's just about plausible that on some platform, for some number type,
+  // performance might be different on zero buffers vs random buffers. But it
+  // shouldn't matter that we recreate the random engine every time, getting
+  // the same random values again.
+  iree_uk_write_random_buffer(rowmajor_lhs_buffer, rowmajor_lhs_buffer_size,
+                              lhs_type, engine);
+  iree_uk_write_random_buffer(rowmajor_rhs_buffer, rowmajor_rhs_buffer_size,
+                              rhs_type, engine);
+  iree_uk_write_random_buffer(rowmajor_init_out_buffer,
+                              rowmajor_out_buffer_size, out_type, engine);
+  iree_uk_write_random_buffer(rowmajor_out_buffer, rowmajor_out_buffer_size,
+                              out_type, engine);
+  iree_uk_write_random_buffer(packed_lhs_buffer, packed_lhs_buffer_size,
+                              lhs_type, engine);
+  iree_uk_write_random_buffer(packed_rhs_buffer, packed_rhs_buffer_size,
+                              rhs_type, engine);
+  iree_uk_write_random_buffer(packed_out_buffer, packed_out_buffer_size,
+                              out_type, engine);
+  mmt4d_params.lhs_buffer = packed_lhs_buffer;
+  mmt4d_params.rhs_buffer = packed_rhs_buffer;
+  mmt4d_params.out_buffer = packed_out_buffer;
+  pack_lhs_params.in_buffer = rowmajor_lhs_buffer;
+  pack_lhs_params.out_buffer = packed_lhs_buffer;
+  pack_rhs_params.in_buffer = rowmajor_rhs_buffer;
+  pack_rhs_params.out_buffer = packed_rhs_buffer;
+  pack_out_params.in_buffer = rowmajor_init_out_buffer;
+  pack_out_params.out_buffer = packed_out_buffer;
+  unpack_out_params.in_buffer = packed_out_buffer;
+  unpack_out_params.out_buffer = rowmajor_out_buffer;
+
+  void* rowmajor_reference_out_buffer = malloc(rowmajor_out_buffer_size);
+  memcpy(rowmajor_reference_out_buffer, rowmajor_init_out_buffer,
+         rowmajor_out_buffer_size);
+  iree_uk_reference_rowmajor_matmul(
+      params->type, params->M, params->K, params->N, rowmajor_lhs_buffer,
+      rowmajor_rhs_buffer, rowmajor_reference_out_buffer, FLAG_accumulate);
+
+  int64_t batch_count = 1;
+  int64_t total_iterations = 0;
+  bool already_checked_results = false;
+  while (iree_benchmark_keep_running(benchmark_state, batch_count)) {
+    for (int i = 0; i < batch_count; ++i) {
+      iree_uk_pack(&pack_lhs_params);
+      iree_uk_pack(&pack_rhs_params);
+      if (FLAG_accumulate) iree_uk_pack(&pack_out_params);
+      iree_uk_mmt4d(&mmt4d_params);
+      iree_uk_unpack(&unpack_out_params);
+      if (!already_checked_results) {
```

Comment:
Shouldn't this be done outside the benchmarking loop?

Reply:
Good point, done.